src/swarm/node/request/RequestStats.d

```diff
@@ -745,4 +745,327 @@ public class RequestStats
+        private struct ClientRequest
+        {
+            /// Scheduled-for-removal request that was sent.
+            cstring request;
+
+            /// Name of the client that sent the request.
+            cstring client;
+
+            /*******************************************************************
+
+                Struct hashing function. Required for use as an AA key in D1 and
+                D2.
+
+                Returns:
+                    hash of this instance
+
+            *******************************************************************/
+
+            public hash_t toHash ( ) /* d1to2fix_inject: const */
+            {
+                return Fnv1a.combined(this.request, this.client);
+            }
+
+            /*******************************************************************
+
+                Struct equality function. Required for use as an AA key in D2.
+
+                Params:
+                    rhs = other instance to compare against
+
+                Returns:
+                    true if rhs is equal to this instance
+
+            *******************************************************************/
+
+            mixin (genOpEquals("
+            {
+                return this.request == rhs.request && this.client == rhs.client;
+            }
+            "));
+
+            /*******************************************************************
+
+                Struct comparison function. Required for use as an AA key in D1.
+
+                TODO: can be removed, when we ditch D1.
+
+                Params:
+                    rhs = other instance to compare against
+
+                Returns:
+                    0 if rhs is equal to this instance, 1 otherwise
+
+            *******************************************************************/
+
+            mixin (genOpCmp("
+            {
+                return this.request != rhs.request || this.client != rhs.client;
+            }
+            "));
+        }
+
+        /// Per-client/request information map.
+        private ClientRequestInfo[ClientRequest] client_request_info;
+
+        /***********************************************************************
+
+            Called when a scheduled-for-removal request starts to be handled.
+
+            Params:
+                rq = name of request
+                client = name of client that sent the request
+                addr = remote addr/port of client
+
+        ***********************************************************************/
+
+        public void started ( cstring rq, cstring client, AddrPort addr )
+        {
+            auto now = time(null);
+            auto cl_rq = ClientRequest(client, rq);
+            auto info = cl_rq in this.client_request_info;
+            if ( info is null )
+            {
+                this.client_request_info[ClientRequest(client.dup, rq.dup)]
+                    = ClientRequestInfo.init;
```

Comment:
I was worried that this opens an attack window, but the allocation happens only once per client per request :+1:.

Reply:
Right.